SO-2902: Add multi-language RF2 import support.

zoliMolnar opened the pull request:
This pull request enables multi-language RF2 importing.

Commit:
[rf2] Added tests for multi-language rf2 importer.

Commit:
[rf2] Changed description, languagerefset file property to collection.

Commit:
[rf2] Changed logic so it can work with collections.

Commit:
[rf2] Added test import rf2 archives

Commit:
[rf2] Added statusCode check for the concept id, renamed variables to

match what they represent, members ids instead of concept ids

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java:
```diff
@@ -203,5 +207,52 @@ public void import11ExtensionConceptWithVersion() {
+	public void importMoreThanOneLanguageCodeDescription() {
+		final String enDescriptionId = "41320138114";
+		final String svDescriptionId = "24688171113";
+		final String conceptIdODescription = "301795004";
```
Typo

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java:
```diff
@@ -203,5 +207,52 @@ public void import11ExtensionConceptWithVersion() {
-
+	
+	@Test
+	public void importMoreThanOneLanguageCodeDescription() {
```
Please add a number to the test method's name like the others above. As you can see there is an annotation at the top: @FixMethodOrder(MethodSorters.NAME_ASCENDING) which enforces method execution order.

nagyo replied:
importDescriptionsWithMultipleLanguageCodes or importDescriptionsWithDifferentLanguageCodes

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java:
```diff
@@ -203,5 +207,52 @@ public void import11ExtensionConceptWithVersion() {
+	
+	@Test
+	public void importFromMoreThanOneLanguageRefset() {
+		final String enLanguageRefsetConceptId = "34d07985-48a0-41e7-b6ec-b28e6b00adfc";
```
You should decide to either merge the two test cases together (since the RF2 files are the same and you always import both descriptions and language refset members as well) or split the RF2 files and test language refset and description import separately.

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java:
```diff
@@ -203,5 +207,52 @@ public void import11ExtensionConceptWithVersion() {
+	}
+	
+	@Test
+	public void importFromMoreThanOneLanguageRefset() {
```
You are not importing more than one language refset! It should be rather something like:
importLanguageRefsetMembersWithMultipleLanguageCodes or importLanguageRefsetMembersWithDifferentLanguageCodes

nagyo commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java:
```diff
@@ -107,13 +107,20 @@ public void execute(final CommandInterpreter interpreter) {
-			config.setDescriptionsFile(new File(archiveFileSet.getFileName(zipFiles, ReleaseComponentType.DESCRIPTION, contentSubType)));
-			config.setLanguageRefSetFile(languageRefSetFile);
+			final Collection<String> allFileName = archiveFileSet.getAllFileName(zipFiles, ReleaseComponentType.DESCRIPTION, contentSubType);
+			final Collection<File> descriptionFileNames = Collections.emptySet();
```
Please fix the initialization of this set. You create an immutable set and then try to add elements to it??
Here is the javadoc of Collections.emptySet(): "Returns an empty set (immutable)." Javadoc is your friend! Nevertheless this could be solved with Java 8 streams in one line.

nagyo commented on snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java:
```diff
@@ -131,14 +134,18 @@ public void setRelationshipsFile(final File relationshipsFile) {
-
+	
+	public void addLanguageRefSetFiles(final File languageRefSetFile) {
+		this.languageRefSetFiles.add(languageRefSetFile);
```
What happens if I call this add method right after I called setLanguageRefSetFiles(Collection<File> files) ????

nagyo commented on snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java:
```diff
@@ -131,14 +134,18 @@ public void setRelationshipsFile(final File relationshipsFile) {
 	}
-
+	
+	public void addLanguageRefSetFiles(final File languageRefSetFile) {
```
If I would like to add **one** language refset file to a collection why the method name addLanguageRefSetFile_**s**_?

nagyo commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java:
```diff
@@ -107,13 +107,20 @@ public void execute(final CommandInterpreter interpreter) {
+			for (String fileName : allFileName) {
+				descriptionFileNames.add(new File(fileName));
+			}
+			config.setDescriptionsFiles(descriptionFileNames);
```
We are in a for loop here, why do you add all the description files to the config in each iteration?

nagyo commented on snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java:
```diff
@@ -58,10 +61,10 @@
 	private File rootFile;
 	private File conceptsFile;
-	private File descriptionsFile;
+	private Collection<File> descriptionsFiles = Collections.emptySet();
```
Please fix the initialization of this collection

nagyo replied:
Thanks to this the collection will be initialized still as immutable.

nagyo commented on snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java:
```diff
@@ -58,10 +61,10 @@
 	private File relationshipsFile;
 	private File statedRelationshipsFile;
-	private File languageRefSetFile;
+	private Collection<File> languageRefSetFiles = Collections.emptySet();
```
Please fix the initialization of this collection

nagyo commented on snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java:
```diff
@@ -18,13 +18,16 @@
 import java.io.File;
```
I get the following exception when I open the import wizard in the client (server-client mode): 
`java.lang.IllegalArgumentException: Could not find property with name descriptionsFile in class class com.b2international.snowowl.snomed.importer.net4j.ImportConfiguration
	at org.eclipse.core.internal.databinding.beans.BeanPropertyHelper.getPropertyDescriptor(BeanPropertyHelper.java:168)
	at org.eclipse.core.databinding.beans.PojoProperties.value(PojoProperties.java:119)
	at org.eclipse.core.databinding.beans.PojoProperties.value(PojoProperties.java:89)
	at org.eclipse.core.databinding.beans.PojoObservables.observeValue(PojoObservables.java:76)
	at org.eclipse.core.databinding.beans.PojoObservables.observeValue(PojoObservables.java:56)
	at com.b2international.snowowl.snomed.importer.ui.wizard.page.ImportDetailsWizardPage.bindTextField(ImportDetailsWizardPage.java:319)
	at com.b2international.snowowl.snomed.importer.ui.wizard.page.ImportDetailsWizardPage.initDataBinding(ImportDetailsWizardPage.java:236)
	at com.b2international.snowowl.snomed.importer.ui.wizard.page.AbstractImportWizardPage.createControl(AbstractImportWizardPage.java:67)`

nagyo commented on snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/SnomedImportRequest.java:
```diff
@@ -69,16 +70,23 @@ protected void requesting(final ExtendedDataOutputStream out, final OMMonitor mo
 			monitor.worked(); // 1
 			
+			for (File descfile : descriptionsFiles) {
+				writeComponent(out, descfile, monitor);
```
This is not going to work like this, see how refset URLs or excluded refset ids are sent over the wire.

nagyo commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/net4j/SnomedImportIndication.java:
```diff
@@ -92,12 +93,12 @@ protected void indicating(final ExtendedDataInputStream in, final OMMonitor moni
 			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setConceptsFile(f); }}, monitor.fork());
-			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setDescriptionsFile(f); }}, monitor.fork());
+			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setDescriptionsFiles(ImmutableList.of(f)); }}, monitor.fork());
 			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setTextDefinitionFile(f); }}, monitor.fork());
```
This is not going to work like this, see how refset URLs or excluded refset ids are sent over the wire.

nagyo commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/SnomedRefSetNameCollector.java:
```diff
@@ -173,11 +173,15 @@ public void handleRecord(int recordCount, List<String> record) {
-			CsvParser parser = new CsvParser(descriptionReader, getFileName(url), CSV_SETTINGS, descriptionParserCallback, DESCRIPTION_FIELD_COUNT);
-			parser.parse();
+			if (!configuration.getDescriptionsFiles().isEmpty()) {
+				for(File descFile : configuration.getDescriptionsFiles()) {
```
missing space

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java:
```diff
@@ -203,5 +207,52 @@ public void import11ExtensionConceptWithVersion() {
+		importArchive("SnomedCT_Release_INT_20150131_new_concept_for_languageCode.zip");
+		getComponent(branchPath, SnomedComponentType.CONCEPT, conceptIdODescription).statusCode(200);
+		
+		importArchive("SnomedCT_Release_INT_20150201_new_description_with2_language_code.zip", importConfiguration);
```
The description in sct2_Description_Delta-sv_INT_20150201 has a language code 'en', please fix that!

Commit:
[rf2] Changed data binding properties.

Commit:
[rf2] Changed loops to streams moved adding description and language...

refset files logic outside of loop

Commit:
[rf2] Merged the two test cases into one.

Commit:
[rf2] Changed logic of reading and writing of description files and...

language refset files for net4j indications.

Commit:
[rf2] Added final modifiers.

Commit:
[rf2] Fixed test import file changed language code and refset id.

Commit:
[rf2] Fixed typo.

Commit:
[rf2] Fixed conceptId of refset in the import file.

Commit:
[rf2] Added placeholder properties for databinding.

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java:
```diff
@@ -203,5 +207,36 @@ public void import11ExtensionConceptWithVersion() {
+		final String svLanguageRefsetMemberId = "34d07985-48a0-41e7-b6ec-b28e6b00adfb";
+		final String conceptIdOfDescription = "301795004";
+		
+		
```
Unnecessary extra line

nagyo commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java:
```diff
@@ -93,27 +93,40 @@ public void execute(final CommandInterpreter interpreter) {
 	
 			try {
-				provider.parse(config.toURL(config.getLanguageRefSetFile()));
+				for (File langFile : config.getLanguageRefSetFiles()) {
```
You are already iterating through the available language refset files in the outer for loop, I don't think this is necessary here.

nagyo commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java:
```diff
@@ -93,27 +93,40 @@ public void execute(final CommandInterpreter interpreter) {
+				.getAllFileName(zipFiles, ReleaseComponentType.DESCRIPTION, contentSubType)
+				.stream()
+				.map(fileName -> new File(fileName))
+				.collect(Collectors.toSet());
```
![](https://ci.memecdn.com/8777838.jpg)

nagyo commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/SnomedRefSetNameCollector.java:
```diff
@@ -173,11 +173,15 @@ public void handleRecord(int recordCount, List<String> record) {
+					descriptionReader = new InputStreamReader(url.openStream());
+					final CsvParser parser = new CsvParser(descriptionReader, getFileName(url), CSV_SETTINGS, descriptionParserCallback, DESCRIPTION_FIELD_COUNT);
+					parser.parse();
+					
```
Extra line

nagyo commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/RF2ReleaseRefSetFileCollector.java:
```diff
@@ -67,9 +67,10 @@
+			for (File langFiles : configuration.getLanguageRefSetFiles()) {
+				collectedUrlSet.add(configuration.toURL(langFiles));
 			}
+		
```
Extra line

nagyo commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/validation/SnomedValidationContext.java:
```diff
@@ -162,9 +164,12 @@ private void addRefSetFilesForValidating() throws IOException {
+				if (isValidReleaseFile(langFile)) {
+					releaseFileValidators.add(new SnomedLanguageRefSetValidator(configuration, configuration.toURL(langFile), this));
+				}
+				
```
Extra line

nagyo commented on snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/SnomedImportRequest.java:
```diff
@@ -70,15 +71,27 @@ protected void requesting(final ExtendedDataOutputStream out, final OMMonitor mo
 			
+			final Collection<File> descriptionsFiles = importConfiguration.getDescriptionsFiles();
+			final Collection<File> languageRefSetFiles = importConfiguration.getLanguageRefSetFiles();
 			monitor.worked(); // 1
```
Why is this here? Have you checked how the monitor's work is calculated here?

nagyo commented on snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/SnomedImportRequest.java:
```diff
@@ -70,15 +71,27 @@ protected void requesting(final ExtendedDataOutputStream out, final OMMonitor mo
 			writeComponent(out, importConfiguration.getRelationshipsFile(), monitor.fork());
 			writeComponent(out, importConfiguration.getStatedRelationshipsFile(), monitor.fork());
 			writeComponent(out, importConfiguration.getDescriptionType(), monitor.fork());
-			writeComponent(out, importConfiguration.getLanguageRefSetFile(), monitor.fork()); // + 7
```
That comment is a clue!

nagyo commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/net4j/SnomedImportIndication.java:
```diff
@@ -87,17 +89,44 @@ protected void indicating(final ExtendedDataInputStream in, final OMMonitor moni
 			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setStatedRelationshipsFile(f); }}, monitor.fork());
 			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setDescriptionType(f); }}, monitor.fork());
-			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setLanguageRefSetFile(f); }}, monitor.fork());
+			
```
Extra line

nagyo reviewed (changes requested):
I just realized that multiple Text Definition files could also be present in RF2 archives based on the language code... Please add that feature as well. They work just as the same as description files do.

Commit:
[rf2] Added validation for definition text files, removed extra lines.

Commit:
[rf2] Added proper monitor progress. Added logic to write...

textDefinitionFiles

Commit:
[rf2] Added method to read collection of files and to monitor progress.

Commit:
[rf2] Removed extra line.

Commit:
[rf2] Removed extra line. Added logic to handle multiple...

text definition files.

Commit:
[rf2] Removed unneeded iteration.

Commit:
[rf2] Added logic to handle collection of text definition files.

Commit:
[rf2] Removed extra line.

nagyo committed:
Merge remote-tracking branch 'origin/6.x' into
issue/rf2-multi-language-import-fix

Conflicts:
	snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java

nagyo committed:
SO-2902: Simplify test case

https://snowowl.atlassian.net/browse/SO-2902

nagyo committed:
[snomed.import] Refactor ImportConfiguration

- remove unnecessary fields from import config
- refactored SnomedRefSetNameCollector:
  - do not parse description files for each refset URL, parse them only
once
  - "prepare" resolvable refset labels for the client
- do not send language refset files twice over the wire

nagyo committed:
[snomed.import] Fix cloning of stated relationship file

nagyo committed:
Merge remote-tracking branch 'origin/6.x' into issue/rf2-multi-language-import-fix

apeteri commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java:
```diff
@@ -93,52 +94,56 @@ public void execute(final CommandInterpreter interpreter) {
+		config.setSourceKind(ImportSourceKind.ARCHIVE);
+		config.setArchiveFile(archiveFile);
+		
+		for (final Entry<String, String> label : provider.getRefsetIdToLabelMap().entrySet()) {
```
This label map could also be extended with information from the store (from MAIN, most likely), similar to how it's done in the import wizard.

apeteri replied:
Sorry, I see that the name collector does this already. ✔️

apeteri commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/SnomedRefSetNameCollector.java:
```diff
@@ -67,210 +73,219 @@
-			put(ErroneousAustralianReleaseFileNames.ERRONEOUS_AU_20110531_SNOMED_LANGUAGE_REFSET_NAME, Concepts.REFSET_LANGUAGE_TYPE).
-			put(ErroneousAustralianReleaseFileNames.ERRONEOUS_AU_20120229_SNOMED_LANGUAGE_REFSET_NAME, Concepts.REFSET_LANGUAGE_TYPE).build();
+	private static final String UNLABELED_REFSET_SUFFIX = " (unresolved)";
+	private static final String ESCAPED_UNLABELED_REFSET_SUFFIX = CharMatcher.is(')')
```
`Pattern#quote(String)` will return a String that is treated as a "literal pattern" in regular expressions; suggesting to use that instead of targeted replacements.

apeteri commented on snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/SnomedRefSetNameCollector.java:
```diff
@@ -67,210 +73,219 @@
-						refSetTypeIconId = getRefSetTypeIconIdFromHeader(url, record);
-					} else {					
-						unlabeledRefSetIds.add(record.get(CONCEPT_ID_COLUMN));
+					InputStreamReader descriptionReader = null;
```
`descriptionReader` should get the try-with-resources treatment:

```java

final URL url = configuration.toURL(descFile);
try (final InputStreamReader descriptionReader = new InputStreamReader(url.openStream())) {
    ...
}
```

apeteri reviewed (changes requested):
More copyright dates to move forward:

* `SnomedImportApiTest`
* `ImportRefSetCommand`
* `ListLanguageRefSetsCommand`
* `ImportUtil`
* `RF2ReleaseRefSetFileCollector`
* `SnomedRefSetNameCollector`
* `AbstractSnomedValidator`
* `SnomedConceptValidator`
* `SnomedTaxonomyValidator`
* `SnomedValidationContext`
* `ImportConfiguration`
* `ReleaseFileSet`
* `ReleaseFileSetSelectors`

nagyo committed:
SO-2902: Update license header

https://snowowl.atlassian.net/browse/SO-2902

nagyo committed:
SO-2902: Use Pattern.quote instead of Charmatcher.replace

https://snowowl.atlassian.net/browse/SO-2902

nagyo committed:
SO-2902: Use try-with-resource for description readers

https://snowowl.atlassian.net/browse/SO-2902

nagyo committed:
SO-2902: Fix NPE

https://snowowl.atlassian.net/browse/SO-2902

cmark reviewed (approved):
[![LGTM](https://www.lgtm.in/p/YO8oFNJLs)](https://www.lgtm.in/i/YO8oFNJLs)
